elide-core/src/main/java/com/yahoo/elide/graphql/PersistentResourceFetcher.java

```diff
@@ -263,12 +305,321 @@ private Object fetchRelationship(Environment context) {
         }
     }
 
-    /** stub code **/
-    private Object replaceObjects(Environment context) { return null; }
+    /**
+     * handle UPSERT operation
+     * @param context environment encapsulating graphQL's request environment
+     * @return list of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertObjects(Environment context) {
+        /* sanity check for id and data argument w UPSERT */
+        if(context.ids.isPresent()) {
+            throw new BadRequestException("UPSERT must not include ids");
+        }
+
+        if(!context.data.isPresent()) {
+            throw new BadRequestException("UPSERT must include data argument");
+        }
+
+        Set upsertOps = new HashSet();
+        /* upsert */
+        for(Map<String, Object> input : context.data.get()) {
+            upsertOps.addAll(upsertGraphWalker(input, this::upsertObject, Optional.ofNullable(context.parentResource),
+                    context.requestScope, context.field.getName()));
+        }
+
+        /* fixup relationships */
+        for(Map<String, Object> input : context.data.get()) {
+            upsertGraphWalker(input, this::updateRelationship, Optional.ofNullable(context.parentResource),
+                    context.requestScope, context.field.getName());
+        }
+
+        return upsertOps;
+    }
+
+    /**
+     * utility class to support storing a triplet on each node of the queue used while upserting
+     */
+    private class Triplet {
+        private List<Map<String, Object>> input;
+        private Optional<PersistentResource> parentResource;
+        private String fieldName;
+
+        /**
+         * class constructor
+         * @param input data input
+         * @param parentResource parent
+         * @param fieldName field name
+         */
+        private Triplet(List<Map<String, Object>> input, Optional<PersistentResource> parentResource, String fieldName) {
+            this.input = input;
+            this.parentResource = parentResource;
+            this.fieldName = fieldName;
+        }
+    }
+
+    /**
+     * A function to handle upserting (update/create) objects.
+     */
+    @FunctionalInterface
+    private interface UpsertInterface<T> {
+        T execute(Map<String, Object> input, String id,
+                                   RequestScope requestScope, Class<?> entityClass,
+                                   String fieldName, Optional<PersistentResource> parent);
+    }
+
+    /**
+     * Forms the graph from data {@param input} and upsert's {@param function} all the nodes
+     * @param input input data
+     * @param function function to process nodes
+     * @param parent parent resource
+     * @param requestScope request scope
+     * @param fieldName graphql field name
+     * @return set of {@link PersistentResource} objects
+     */
+    private Set<PersistentResource> upsertGraphWalker(Map<String, Object> input, UpsertInterface function,
```

Comment:
Let's rename the function to just graphWalker.  The function param type should also be a different name.